WidgetDriver: add to_device support

toger5 opened the pull request:
Updated version of: https://github.com/matrix-org/matrix-rust-sdk/pull/4155

This introduces widget driver support for sending to devices messages: "A widget (element call) can request send an action via postmessage to the client (element X). This PRs job is to do everything from parsing the json to sending the to_device cs api request to sending back the response."


<!-- description of the changes in this PR -->

- [ ] Public API changes documented in changelogs (optional)

<!-- Sign-off, if not part of the commits -->
<!-- See CONTRIBUTING.md if you don't know what this is -->
Signed-off-by:

Hywan commented:
Hello,

The PR title contains WIP but a review is requested. The PR description contains almost nothing. Please, move the PR as a draft. Ping me when it's ready, I'll be delighted to review it then.

Hywan commented on crates/matrix-sdk/src/widget/matrix.rs:
```diff
@@ -204,19 +222,170 @@ impl MatrixDriver {
+                    .ok()
+                }}));
+            // The first flatten takes the iterator of Option<Device>'s iterators and converts it to just a iterator over Option<(Device, data)>'s.
+            // The second flatten takes the the iterator over Option<(Device,data)>'s and converts it to just a iterator over Device
```
So the “data” are computed and dropped?

Hywan commented on crates/matrix-sdk/src/widget/matrix.rs:
```diff
@@ -204,19 +222,170 @@ impl MatrixDriver {
+        // We first want to get all missing session before we start any to device sending!
+        client.claim_one_time_keys(messages.iter().map(|(u, _)| u.as_ref())).await?;
+
+        let request = match encrypted {
+            true => {
```
```diff
let request = if encrypted {
    …
} else {
    …
};
```

Hywan commented on crates/matrix-sdk/src/widget/matrix.rs:
```diff
@@ -204,19 +222,170 @@ impl MatrixDriver {
+        };
+
+        let response = client.send(request).await;
+        // if let Ok(res){
+        //     client.mark_request_as_sent(request.request_id(), &res).await;
+        // };
```
Clean up.

Hywan commented on crates/matrix-sdk/src/widget/matrix.rs:
```diff
@@ -204,19 +222,170 @@ impl MatrixDriver {
-pub(crate) struct EventReceiver {
-    rx: UnboundedReceiver<Raw<AnyTimelineEvent>>,
-    _drop_guards: [EventHandlerDropGuard; 2],
+
```
Remove the empty line.

Hywan commented on crates/matrix-sdk/src/widget/mod.rs:
```diff
@@ -269,16 +291,42 @@ impl WidgetDriver {
+                    self.to_device_event_forwarding_guard = Some(guard);
+                    let (mut matrix, events_tx) =
+                        (matrix_driver.to_device_events(), incoming_msg_tx.clone());
+                    tokio::spawn(async move {
```
Please use `matrix_sdk_common::executor::spawn` instead.

toger5 replied:
We need to be careful here for wasm since it will drop the handle and consequently stop the async process.
So this would need a `forget` call on the handle or keeping a reference to the handle while the process should be running (tokio and the wasm spawn behave different in that regard)

Hywan replied:
So please update `matrix_sdk_common::executor::spawn` to do what you need, but don't use `tokio::spawn` directly here.

Hywan commented on crates/matrix-sdk/src/widget/capabilities.rs:
```diff
@@ -62,39 +65,37 @@ pub struct Capabilities {
+        let filter_in = match raw.deserialize_as::<MatrixEventFilterInputData>() {
+            Ok(filter) => MatrixEventFilterInput::Timeline(filter),
+            Err(err) => {
+                error!("Failed to deserialize raw event as MatrixEventFilterInput: {err}");
```
Should be rebased onto the first commit.

Hywan commented on crates/matrix-sdk/src/widget/capabilities.rs:
```diff
@@ -62,39 +65,37 @@ pub struct Capabilities {
-const SEND_TODEVICE: &str = "org.matrix.msc3819.send.to_device";
-const RECEIVE_TODEVICE: &str = "org.matrix.msc3819.receive.to_device";
-const REQUIRES_CLIENT: &str = "io.element.requires_client";
+pub(super) const SEND_EVENT: &str = "org.matrix.msc2762.send.event";
+pub(super) const RECEIVE_EVENT: &str = "org.matrix.msc2762.receive.event";
+pub(super) const SEND_STATE: &str = "org.matrix.msc2762.send.state_event";
+pub(super) const RECEIVE_STATE: &str = "org.matrix.msc2762.receive.state_event";
+pub(super) const SEND_TODEVICE: &str = "org.matrix.msc3819.send.to_device";
+pub(super) const RECEIVE_TODEVICE: &str = "org.matrix.msc3819.receive.to_device";
```
Same here and everywhere else.

Hywan reviewed (changes requested):
Thanks for your work, but I don't think we are there yet. This is hard to review.

I did review up to the 5th patches.

Please, rewrite the history of this pull request. The first commit is super large and, because it has no description, it's quite hard to know what it does. The code is not completely documented. The code is not formatted. Some parts of the code are marked as dead. Some parts of the code are commented.

Maybe the next commits fix that up. In this case, these commits should be merged together.

Also please, add commit descriptions! A good clue is: if a commit does too many things, it means the commit must be split into smaller ones. Also, if providing a description for a commit is complicated, then imagine how hard it is for the reviewer to check the diff.

Hywan commented on crates/matrix-sdk/src/widget/machine/from_widget.rs:
```diff
@@ -237,3 +240,16 @@ impl From<update_delayed_event::unstable::Response> for UpdateDelayedEventRespon
+/// which derives Serialize. (The response struct from Ruma does not derive
+/// serialize)
+#[derive(Serialize, Debug)]
+pub(crate) struct SendToDeviceEventResponse {}
```
Can't we update the tests? What's the impact of changing that?
If you can't change, okay, but please indicate this information in the documentation.

Hywan reviewed (changes requested):
I'm still struggling to review the essence of the PR because there is too much noise to fix beforehand.

Your patches are still a bit big but I can live with it _if_ you add a description to each commit explaining what it does and why. Thanks :-).

toger5 commented on crates/matrix-sdk/src/widget/filter.rs:
```diff
@@ -103,271 +109,382 @@ pub enum StateEventFilter {
+}
+
+// Filter input:
+
```
Seperates the following section which does not implement filters anymore but Filter inputs.
Is there a common style how such seperator is done (or is that considered unhelpful?)

toger5 commented on crates/matrix-sdk/src/widget/filter.rs:
```diff
@@ -103,271 +109,382 @@ pub enum StateEventFilter {
+pub struct FilterInputState {
     #[serde(rename = "type")]
-    pub(super) event_type: TimelineEventType,
+    pub(super) event_type: String,
```
This allow to have a seperate `FilterInputState` vs `FilterInputMessageLike` so that we do not have the single data type that contained state key and msgtype.
(the distinction between state and messageLike was done via event_type before which is easy to misconfigure when using `into()` especially with unknown even types from widgets. Now the event_type is a string and the distinction is done explicitly by Filter::...)

toger5 replied:
I updated it now to be: The correct event type even though its already definde by the Filter enum.
`StateEventType`, `ToDeviceEventType`, `MessageLikeEventType`

toger5 commented on crates/matrix-sdk/src/widget/filter.rs:
```diff
@@ -103,271 +109,382 @@ pub enum StateEventFilter {
+    type Error = serde_json::Error;
+    fn try_from(raw_event: Raw<AnyToDeviceEvent>) -> Result<Self, Self::Error> {
+        // make sure Raw actually was a message like event
+        raw_event.deserialize()?;
+        raw_event.deserialize_as::<FilterInputToDevice>().map(FilterInput::ToDevice)
```
Here this first one is actually unnecassary. A to device message is already valid with only a `type` field so the second implies that the first is working.

toger5 commented on crates/matrix-sdk/src/widget/matrix.rs:
```diff
@@ -204,19 +212,194 @@ impl MatrixDriver {
+        let to_device_handle = self.room.client().add_event_handler(
+            move |raw: Raw<AnyToDeviceEvent>, encryption_info: Option<EncryptionInfo>| {
+                // Deserialize the Raw<AnyToDeviceEvent> to a mutable structure
+                let mut event_with_encryption_flag: Value =
+                    raw.deserialize_as().expect("Invalid event JSON");
+
+                if let Some(content) = event_with_encryption_flag.get_mut("content") {
+                    content["encrypted"] = json!(encryption_info.is_some());
+                }
+                let ev_for_widget = match Raw::<AnyToDeviceEvent>::from_json_string(
+                    event_with_encryption_flag.to_string(),
+                ) {
+                    Ok(ev) => ev,
+                    Err(_) => raw,
+                };
```
I ended up refactoring this to be equivalent to what we do for adding the `room_id`
This has the advantage that we deserialize as: `BTreeMap<String, Box<RawJsonValue>` which should be quicker considering `RawJsonValue` is a trivial deserialization.

toger5 commented:
@Hywan I addressed the more obvious changes and added more details to the larger commits: https://github.com/matrix-org/matrix-rust-sdk/pull/4856/commits/f091aac2961e6c470e26e3e2ccc55f40359a62d2

toger5 commented:
Just `rebased` on main

BillCarsonFr commented on crates/matrix-sdk/src/widget/matrix.rs:
```diff
@@ -208,25 +235,214 @@ impl MatrixDriver {
+
+                async move {
+                    if !device.is_cross_signed_by_owner() {
+                        info!("Device {} is not verified, skipping encryption", device.device_id());
```
`verified` is not the correct term, the device is insecure.
There are more checks to do here, I wonder if we should manage that in `encrypt_event_raw` with some specific setting.

BillCarsonFr commented on crates/matrix-sdk/src/widget/matrix.rs:
```diff
@@ -208,25 +235,214 @@ impl MatrixDriver {
+                    let user_id = user_id.to_owned();
+                    let event_type = event_type.clone();
+                    async move {
+                        let Ok(user_devices) = client.encryption().get_user_devices(&user_id).await else {
```
We shuold probably handle that in the crypto crate as per https://github.com/matrix-org/matrix-rust-sdk/issues/4147

BillCarsonFr commented on crates/matrix-sdk/src/widget/matrix.rs:
```diff
@@ -208,25 +235,214 @@ impl MatrixDriver {
+        }
+
+        let request = if encrypted {
+            // We first want to get all missing session before we start any to device
```
actually you should do that before encrypting.

BillCarsonFr commented on crates/matrix-sdk/src/widget/matrix.rs:
```diff
@@ -208,25 +235,214 @@ impl MatrixDriver {
+    }
+}
+
+fn with_attached_encryption_flag(
```
why are we doing that?

BillCarsonFr commented on crates/matrix-sdk/tests/integration/widget.rs:
```diff
@@ -417,32 +419,63 @@ async fn test_receive_live_events() {
         .await;
 
-    let msg = recv_message(&driver_handle).await;
-    assert_eq!(msg["api"], "toWidget");
```
Sounds like something for insta:assert_json_snapshot ?

toger5 commented:
Lets close this since we have a plan how to restructure this into new PR's.
